### ISSUE TITLE
Document DB Data Format

### ISSUE THREAD

**Issue by nibanks** [microsoft/netperf#16]
I was taking a look at an example output [here](https://github.com/microsoft/netperf/blob/deploy/data/secnetperf/2023-12-04-15-16-48/test_result.json), and I have a few suggestions. Also, we need to not forget to update [api-interface-schema.md](https://github.com/microsoft/netperf/blob/main/docs/api-interface-schema.md) with the format once we're done.

## TestRuns

```json
  "TestRuns": {
    "Max throughput test using TCP protocol with -upload:10000, -timed:1": [
      "Started!",
      "",
      "Result: 1172439040 bytes @ 934493 kbps (10037.006 ms).",
      "App Main returning status 0"
    ],
    "Max throughput test using QUIC protocol with -upload:10000, -timed:1": [
      "Started!",
      "",
      "Result: 1264713728 bytes @ 908461 kbps (11137.188 ms).",
      "App Main returning status 0"
    ]
  },
```

- The file doesn't need to be human readable on its own. We will document the schema and tools will have documentation to indicate what certain arguments do, so we don't need the `"Max throughput test using..."` parts. We should replace it with simply the arguments of the tool used.

- This shouldn't be a list of test runs but should instead be a multi-level list of tests, each with a set of multiple runs. I would format it like this:

```json
"Tests": {
  "1": {
    "ServerArgs": "...",
    "ClientArgs": "...",
    "Runs": [
      "Result: 1172439040 bytes @ 934493 kbps (10037.006 ms).",
      "..."
    ]
  }
}
```
- Additionally, you will notice I removed everything but the "Result: " line in the `Runs` list. I'm still thinking if we should further preprocess and simplify the output here. Currently, we have several types of results, depending on what we're looking at/for:

1. For throughput, we only care about the rate number (i.e. in kbps).
2. For RPS, we only care about the RPS rate number.
3. For HPS, we only care about the HPS rate number.
4. For RPS latency, it's more complex because we care about the latency curve, which consists of multiple numbers, likely stored in a separate file; but maybe we store certain percentiles here?

## TestEnv

```json
"TestEnv": {
  "Client": {
    "NIC": "Mellanox ConnectX-5",
    "CPU": "Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz",
    "OS": "Windows Server 2022",
    "Arch": "x64"
  },
  "Server": {
    "NIC": "Mellanox ConnectX-5",
    "CPU": "Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz",
    "OS": "Windows Server 2022",
    "Arch": "x64"
  }
},
"RunDate": "2023-12-04-15-15-27",
"MachineName": "netperf-secnetp",
"TestConfig": {
  "MsQuicCommit": "5a6975e82f6ab97b83d10abf92f862e24910bf03",
  "PerfTool": "secnetperf"
}
```

- Let's put this first in the file.
- `MachineName` should actually be `Client` and `Server` specific and not go in the top-level part.
- Instead of `TestConfig`, let's call it `Tool`, put it first, and rename the fields in it:
```json
"Tool": {
  "Name": "secnetperf",
  "Commit": "5a6975e82f6ab97b83d10abf92f862e24910bf03",
  "Config": "-TLS Schannel"
}
```
- We need the `Config` option to handle various build configuration options for the tool itself.
- For `Client` and `Server, let's put those last. Additionally, for their field values, I think we should try to compress things some, as we really don't need all the explicit details. Maybe something like this:
```json
"Client": {
  "NIC": "CX5",
  "CPU": "8272CL",
  "OS": "WS", # I thought about Windows, Win or WinSer too, but we should eventually support Windows Client as well
  "Ver": "2022"
  "Arch": "x64"
}
```
- Should we capture NIC driver versions? What about any other configuration knobs for them?
- Another thought: Should we really separate out Client and Server here? While it would be nice to test one type against another, we've never done that up until now, and if we do go down that road, the matrix would explode... Food for thought.

## General Issues

- With how this is currently set up, you can only put one set of runs for a single build/OS configuration in a file. In other words, Windows Server 2019 and Windows Server 2022 results go in separate files; Ubuntu 20.04 goes in a different one; and so on. But you name this file `test_result.json`, which is too generic. So, we need to either:
  - update the file to support multiple configurations in the same file, or
  - we need to name the file based on the configuration.
- I think putting stuff like that into the file name gets pretty messy, which we really stuff like OS type and version, TLS type, IO type, etc. So we should probably update the json format to have a top-level array to store per-configuration test/run data.
- **BUT** this gets further complicated by my next thought: I not sure using a directory structure based on date/time is the right approach. But I'm not sure. A few thoughts here:
  - We want to support running a subset of tests when part of a test matrix changes. For instance, if we update the prerelease version of Windows, then run all the Windows tests again on that new build. If MsQuic changes commit, run it on all OS versions.
  - So, how do we store this data, index it and search it appropriately. If the dashboard needs to grab "the latest" data where does it look? If we have files by date, some of the latest might be in one file while the rest is in one or more other files.
- So, I then wonder if we should have just a `raw` folder that has each configurations run, named by date/time and/or guid (for uniqueness), and then another directory structure that can be used as an index to point to the relevant raw files (or just the latest?).
- This is where a real DB might be a better option, but I'm not sure.
- Additionally, should we have a separate directory for the tool name? What if we want to support multiple tools for the same XDP commit (i.e. if we get running MsQuic tests for a given XDP commit; plus we have the normal XDP perf results)?

Bottom line, I think this is on the right track, but we need to discuss some of these things for sure.

**Comment by maolson-msft** [comment]
As discussed in IM, let's use SQL (which if we want to store the data in git means a sqlite file), and document the schema externally rather than embedding friendly strings in the data as Nick mentions above.

**Comment by nibanks** [comment]
Agreed. Just thinking about what tables might be needed in the DB mode.... mostly everything discussed in the issue I opened, but split into separate tables to reduce any duplication.

### Environment Table

- Unique ID
- OS type and version
- Architecture
- NIC type (and version?)
- CPU type

### Machine Table

- Unique ID
- Environment ID
- Name

### Tool Table

- Unique ID
- Name
- Commit
- Build date/time (?)
- Config options (for secnetperf: TLS, kernel, etc.)

### Test Table

- Unique ID
- Tool ID
- Arguments

### Run Table

- Unique ID (actually necessary?)
- Test ID
- Server Machine ID
- Client Machine ID
- Result

Then, as we add new runs we only have to add minimal new rows:

- If it is because of an OS change, add new environment row(s)
- If it is because of a tool change, add new tool row(s)
- Add new machine row if it's a new combination (todo: do we really care to differentiate machine from environment?)
- Add new rows for all the runs

What about latency results? Do we need a whole table for that, or do we stuff all the values in one "results" field?
How do we handle ordering of results in a timeline view? MsQuic does it based on the order of commits, but that's not enough since we need to consider changes in the OS over time as well. Is "Date / Time" simply the way to go and assume when it was executed, it was in order?

**Comment by ProjectsByJackHe** [comment]
To address your points, here are some proposed solutions:

(1): Do we care to differentiate between machine and environment?

So, some physical machines may have some unknown issue that causes throughput / latency data to be off (as you pointed out). Having this distinction of a unique environment setting (class) and a specific machine (instance) is helpful. That would mean though, whenever there is a new OS build, we'd have to append to 2 tables. Also, we have around 60 physical machines, then the Machines table would be filled with rows with duplicate data.

It might be easier to just add "Machine name" to the environments table. 

(2): How do we handle latency results?

I was thinking instead of a single "Results" field in the Runs table, we could have a "Metadata" (string) and "Value" (numeric real) field. In the metadata field, we could specify stuff like "P90 latency" or anything else that's helpful or just NULL if enough context is given by the "Arguments" field in the Test table. (I assume "Arguments" is the powershell command we run for the test). 

That way, it's friendly to query for analytical purposes, although there would be more rows appended.

**Comment by nibanks** [comment]
Keeping this issue around until we have a final format (at least for QUIC), and it's been documented.

**Comment by ProjectsByJackHe** [comment]
Documented here: https://github.com/microsoft/netperf/blob/main/docs/sql.jpg
